### PR TITLE
From workbranch, XML-doc additions, minor setting change to NetMQ.csproj

### PR DESCRIPTION
**Pull request by JamesWHurst**
Aside from filling in more of the XML-documentation (geez that is tedious!), I followed y'all's advice and removed the .chm help-file, since that really does not belong in the repository, and moved the Sandcastle help-file project into the Help directory where it belongs. To support easy generation of the help file, I set the option on the NetMQ project-file to generate a XML-doc output file (this does not impact the normal executable output files) when in DEBUG-mode only, and also the RELEASE mode should not output .pdb files nor TRACE output.

### CODE REVIEW

File: src/NetMQ/NetMQFrame.cs [hunk 1]
```diff
@@ -278,6 +278,33 @@ public override bool Equals(object obj)
         }
 
         /// <summary>
+        /// Return true if this one and the other NetMQFrame are equal, or both are null.
+        /// </summary>
+        /// <param name="one">one frame to compare against the other</param>
+        /// <param name="other">the other frame to compare</param>
+        /// <returns>true if both frames are equal</returns>
+        public static bool operator ==(NetMQFrame one, NetMQFrame other)
```
Comment (leverate-somech) [inline src/NetMQ/NetMQFrame.cs]: I think implementing the == operator is a little problematic.

First it might break existing systems, when you want to make sure it is equals you use .Equals and == for reference equality

File: src/NetMQ/zmq/Transports/Tcp/TcpConnector.cs [hunk 2]
```diff
@@ -180,20 +189,22 @@ private void StartConnecting()
...
+        /// <param name="bytesTransferred">the number of bytes that were transferred</param>
+        /// <exception cref="NetMQException">A non-recoverable socket error occurred.</exception>
         public void OutCompleted(SocketError socketError, int bytesTransferred)
         {
+            m_ioObject.RemoveSocket(m_s);
+            m_isHandleValid = false;
+
             if (socketError != SocketError.Success)
             {
-                m_ioObject.RemoveSocket(m_s);
```
Comment (leverate-somech) [inline src/NetMQ/zmq/Transports/Tcp/TcpConnector.cs]: why did you remove the above code?

File: src/NetMQ/NetMQFrame.cs (hunk 1 above)
Comment (JamesWHurst) [inline src/NetMQ/NetMQFrame.cs]: This was not mine - I merged that in from your current master branch, trying to minimize your merging work. I've no idea why this was added.

File: src/NetMQ/zmq/Transports/Tcp/TcpConnector.cs (hunk 2 above)
Comment (JamesWHurst) [inline src/NetMQ/zmq/Transports/Tcp/TcpConnector.cs]: I don't remember doing that. Did I err in trying to merge from the master, before committing? If not - then this is an error. I had no intention of changing that.

Comment (JamesWHurst) [inline src/NetMQ/zmq/Transports/Tcp/TcpConnector.cs]: Just checked - yes, that's what was in your master-branch, which I merged into my edited code before commiting it, attempting to reduce your work in doing this merge.  It doe NOT look like correct code!  Why mark the handle as invalid, unless the socketError is indeed != Success  ?!   How did this happen?

File: src/NetMQ/NetMQFrame.cs (hunk 1 above)
Comment (somdoron) [inline src/NetMQ/NetMQFrame.cs]: yes. It was there before your commit, somehow github show this as new code

**Comment by somdoron** [comment]
@JamesWHurst I review the changes again, looks fine. For next time I think using rebase can make the pull request easier to merge.

Anyway it's still not mergable...

**Comment by JamesWHurst** [comment]
Ok - correction to TcpConnector made, some additional XML-doc commits added, and ought to be ready now for merging. At this point, *all* of the explicitly-thrown exceptions within NetMQ are documented (meaning, every exception that is thrown within a given method, is included within the XML-doc for that method). Also, every exception that is thrown within NetMQ has some kind of debugging information included within it, to make these easier to track down.

**Comment by JamesWHurst** [comment]
I'll wait until the other pull-request is merged in, and this one, before I do any more work on this -- so that I don't duplicate anyone else's work. Please advise if there're still any issues with this PR.

**Comment by JamesWHurst** [comment]
What is the status of this pull-request? Is there still a reason for it not to be merge-able?  I'm sort of in a holding-pattern waiting on this.

**Comment by somdoron** [comment]
@JamesWHurst pull request cannot be merged at the moment because of conflicts, can you fix this?

**Comment by JamesWHurst** [comment]
I can try - the conflicts may be growing with every other merge.  How do we find out what the conflicts are?

**Comment by drewnoakes** [comment]
Fetch the latest NetMQ master to your repo, checkout your PR's branch,
rebase (not merge) on top of the upstream master. Fix conflicts as your
commits replay. Once all looks good, you can force-push back to your fork
repo and the PR will update.

This might be useful:
https://www.atlassian.com/git/tutorials/merging-vs-rebasing

**Comment by JamesWHurst** [comment]
Ok - done.  That, was a lot of work!  Thanks for the instruction.

**Comment by JamesWHurst** [comment]
Note: I enabled code analysis within the NetMQ project, with the rule set to Managed Minimum. I think that's a good setting to have in general, as it helps point out certain coding flaws. It did complain about a class not disposing it's disposable parts. So I fixed that by making several classes implement IDisposable. But I didn't want to add the IDisposable interface just yet, since this PR is almost entirely just comments, so - I did not change *that* as yet.

**Comment by adamralph** [comment]
The build of this PR is hanging. It's currently occupying both Jetbrains agents at Codebetter CI - http://teamcity.codebetter.com/agents.html

Can someone please kill those builds? Other projects are waiting in the queue and are blocked by them.

It would be good if you could also change your build config to fail the build if it goes over a certain time. One of those builds has been hanging for over 22 hours already.

Given the estimated time for your builds is just over 1 minute, the threshold should probably be quite low, e.g. 10 minutes.

**Comment by drewnoakes** [comment]
@adamralph thanks for pointing this out. I don't know who has admin access to this project. Perhaps @somdoron?

@JamesWHurst, hold off on pushing any more commits to this branch as each push will tie up another build agent. Do the tests run ok on your machine?

**Comment by adamralph** [comment]
@ajepst kindly cancelled the two original builds, but that last push has kicked off another one...

**Comment by somdoron** [comment]
@adamralph <https://github.com/adamralph> fixed.

**Comment by adamralph** [comment]
Thank you @somdoron !

**Comment by JamesWHurst** [comment]
Ok - I changed to a different machine, and the tests are running fine.  That's what I get for mixing too many projects together at the same time.  I just did another commit -- and hopefully this is set to pass tests and be ready to merge.

**Comment by drewnoakes** [comment]
I'll take a look at this in the next evening or so. There were a few things
I wanted to ask you about before merging iirc.

[via phone]

**Comment by drewnoakes** [comment]
@JamesWHurst, I see this conflicts again -- don't worry. I'll merge your API doc changes in, hopefully tonight.
